File: apps/android/app/src/main/java/com/fitnessapp/ui/workout/MuscleMapView.kt
package com.fitcoach.app.ui.workout

import android.graphics.Color
import android.webkit.WebView
import android.webkit.WebViewClient
import androidx.compose.runtime.*
import androidx.compose.ui.Modifier
import androidx.compose.ui.viewinterop.AndroidView

/**
 * Renders an inline SVG muscle diagram in a WebView and highlights the
 * given muscles using #A3FF12 (the FitCoach accent colour).
 *
 * On every recomposition where [highlightedMuscles] changes, the existing
 * WebView is updated via evaluateJavascript — no full reload required.
 */
@Composable
fun MuscleMapView(
    highlightedMuscles: List<String>,
    modifier: Modifier = Modifier
) {
    // Keep a stable reference so we can push JS updates after initial load.
    val webViewRef = remember { mutableStateOf<WebView?>(null) }

    // Track whether the page has finished loading so we can safely inject JS.
    val pageReady = remember { mutableStateOf(false) }

    // Rebuild the HTML whenever the highlighted set changes.
    val html = remember(highlightedMuscles) { buildMuscleMapHtml(highlightedMuscles) }

    AndroidView(
        factory = { context ->
            WebView(context).apply {
                setBackgroundColor(Color.parseColor("#0B0C10"))
                settings.javaScriptEnabled = true
                settings.domStorageEnabled = true
                webViewClient = object : WebViewClient() {
                    override fun onPageFinished(view: WebView, url: String) {
                        pageReady.value = true
                    }
                }
                loadDataWithBaseURL(null, html, "text/html", "UTF-8", null)
                webViewRef.value = this
            }
        },
        update = { webView ->
            // On recomposition with new highlights, push a JS update instead
            // of reloading the full HTML document.
            if (pageReady.value) {
                webView.evaluateJavascript(buildUpdateJs(highlightedMuscles), null)
            } else {
                // Page not ready yet — reload with fresh HTML (e.g. first render
                // or after the view is recycled by the Compose runtime).
                pageReady.value = false
                webView.loadDataWithBaseURL(null, html, "text/html", "UTF-8", null)
            }
        },
        modifier = modifier
    )
}

// ---------------------------------------------------------------------------
// HTML / SVG helpers
// ---------------------------------------------------------------------------

private val ALL_MUSCLES = listOf(
    "chest", "shoulders", "biceps", "triceps", "abs",
    "quads", "hamstrings", "glutes", "calves", "lats", "traps", "lower_back"
)

private fun buildMuscleMapHtml(highlighted: List<String>): String {
    val highlightedSet = highlighted.toSet()

    // Build per-muscle style rules.
    val styleRules = ALL_MUSCLES.joinToString("\n") { muscle ->
        val fill = if (muscle in highlightedSet) "#A3FF12" else "#1C1D26"
        "#$muscle { fill: $fill; transition: fill 0.3s ease; }"
    }

    // Simple schematic SVG — each region is a labelled path with its id
    // matching the muscle name used in the API.
    return """
<!DOCTYPE html>
<html>
<head>
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<style>
  html, body {
    margin: 0; padding: 0;
    background: #0B0C10;
    display: flex;
    justify-content: center;
    align-items: center;
    height: 100%;
    overflow: hidden;
  }
  svg {
    width: 100%;
    max-width: 340px;
    height: auto;
  }
  $styleRules
  text {
    fill: #8A8A9A;
    font-family: sans-serif;
    font-size: 11px;
    pointer-events: none;
    user-select: none;
  }
</style>
</head>
<body>
<svg viewBox="0 0 200 420" xmlns="http://www.w3.org/2000/svg">

  <!-- TRAPS -->
  <path id="traps"
    d="M70,30 Q100,20 130,30 L125,55 Q100,45 75,55 Z"/>
  <text x="100" y="43" text-anchor="middle">traps</text>

  <!-- SHOULDERS -->
  <path id="shoulders"
    d="M55,55 Q70,45 75,65 L65,85 Q52,75 50,65 Z
       M145,55 Q130,45 125,65 L135,85 Q148,75 150,65 Z"/>
  <text x="42" y="72" text-anchor="middle">delts</text>

  <!-- CHEST -->
  <path id="chest"
    d="M75,65 Q100,58 125,65 L122,100 Q100,108 78,100 Z"/>
  <text x="100" y="87" text-anchor="middle">chest</text>

  <!-- BICEPS -->
  <path id="biceps"
    d="M55,85 Q63,80 65,100 L60,120 Q52,115 50,100 Z
       M145,85 Q137,80 135,100 L140,120 Q148,115 150,100 Z"/>
  <text x="42" y="105" text-anchor="middle">biceps</text>

  <!-- TRICEPS (back-of-arm schematic shown on side) -->
  <path id="triceps"
    d="M50,100 Q45,115 48,130 L56,125 Q55,112 58,100 Z
       M150,100 Q155,115 152,130 L144,125 Q145,112 142,100 Z"/>
  <text x="158" y="118" text-anchor="middle">tris</text>

  <!-- ABS -->
  <path id="abs"
    d="M78,100 Q100,108 122,100 L120,160 Q100,168 80,160 Z"/>
  <text x="100" y="135" text-anchor="middle">abs</text>

  <!-- LATS -->
  <path id="lats"
    d="M63,85 Q75,95 78,130 L68,145 Q55,130 55,110 Z
       M137,85 Q125,95 122,130 L132,145 Q145,130 145,110 Z"/>
  <text x="44" y="118" text-anchor="middle">lats</text>

  <!-- LOWER BACK -->
  <path id="lower_back"
    d="M80,160 Q100,168 120,160 L118,185 Q100,192 82,185 Z"/>
  <text x="100" y="176" text-anchor="middle">lower back</text>

  <!-- GLUTES -->
  <path id="glutes"
    d="M82,185 Q100,192 118,185 L115,215 Q100,222 85,215 Z"/>
  <text x="100" y="203" text-anchor="middle">glutes</text>

  <!-- QUADS -->
  <path id="quads"
    d="M82,215 Q90,210 98,215 L96,280 Q90,285 82,280 Z
       M118,215 Q110,210 102,215 L104,280 Q110,285 118,280 Z"/>
  <text x="100" y="250" text-anchor="middle">quads</text>

  <!-- HAMSTRINGS -->
  <path id="hamstrings"
    d="M82,280 Q88,278 96,280 L95,340 Q88,345 82,340 Z
       M118,280 Q112,278 104,280 L105,340 Q112,345 118,340 Z"/>
  <text x="100" y="312" text-anchor="middle">hams</text>

  <!-- CALVES -->
  <path id="calves"
    d="M82,340 Q88,338 95,340 L94,390 Q88,395 82,390 Z
       M118,340 Q112,338 105,340 L106,390 Q112,395 118,390 Z"/>
  <text x="100" y="368" text-anchor="middle">calves</text>

</svg>
</body>
</html>
""".trimIndent()
}

/** JavaScript snippet that updates muscle fills without a page reload. */
private fun buildUpdateJs(highlighted: List<String>): String {
    val allMusclesJson = ALL_MUSCLES.joinToString(",") { "\"$it\"" }
    val highlightedJson = highlighted.joinToString(",") { "\"$it\"" }
    return """
(function() {
    var all = [$allMusclesJson];
    var highlighted = [$highlightedJson];
    all.forEach(function(muscle) {
        var el = document.getElementById(muscle);
        if (!el) return;
        el.style.fill = highlighted.indexOf(muscle) !== -1 ? '#A3FF12' : '#1C1D26';
    });
})();
""".trimIndent()
}

File: apps/ios/FitnessCoach/Views/Workout/MuscleMapView.swift
import SwiftUI
import WebKit

// MARK: - MuscleMapSide

enum MuscleMapSide {
    case front
    case back
}

// MARK: - MuscleMapView

struct MuscleMapView: View {
    let highlightedMuscles: [String]
    var side: MuscleMapSide = .front

    @State private var activeSide: MuscleMapSide
    @State private var webViewRef: WKWebView?

    init(highlightedMuscles: [String], side: MuscleMapSide = .front) {
        self.highlightedMuscles = highlightedMuscles
        self.side = side
        _activeSide = State(initialValue: side)
    }

    var body: some View {
        VStack(spacing: 0) {
            MuscleMapWebView(
                side: activeSide,
                highlightedMuscles: highlightedMuscles,
                onWebViewCreated: { wv in webViewRef = wv }
            )
            .frame(maxWidth: .infinity, maxHeight: .infinity)
            .background(Color(hex: "#0B0C10"))

            // Front / Back toggle
            HStack(spacing: 0) {
                toggleButton(label: "Front", targetSide: .front)
                toggleButton(label: "Back",  targetSide: .back)
            }
            .padding(.horizontal, 24)
            .padding(.vertical, 12)
            .background(Color(hex: "#12131A"))
        }
        .background(Color(hex: "#0B0C10"))
    }

    @ViewBuilder
    private func toggleButton(label: String, targetSide: MuscleMapSide) -> some View {
        let isActive = activeSide == targetSide
        Button {
            withAnimation(.easeInOut(duration: 0.2)) {
                activeSide = targetSide
            }
        } label: {
            Text(label)
                .font(.subheadline.weight(.semibold))
                .frame(maxWidth: .infinity)
                .padding(.vertical, 8)
                .background(isActive ? Color(hex: "#A3FF12") : Color.clear)
                .foregroundColor(isActive ? Color(hex: "#0B0C10") : Color(hex: "#F0F0F0").opacity(0.6))
                .clipShape(RoundedRectangle(cornerRadius: 8))
        }
        .animation(.easeInOut(duration: 0.2), value: activeSide)
    }
}

// MARK: - WKWebView UIViewRepresentable

struct MuscleMapWebView: UIViewRepresentable {
    let side: MuscleMapSide
    let highlightedMuscles: [String]
    var onWebViewCreated: ((WKWebView) -> Void)?

    // Colour constants
    private let accentHex   = "#A3FF12"
    private let defaultHex  = "#1C1D26"
    private let bgHex       = "#0B0C10"
    private let labelHex    = "#F0F0F0"

    func makeUIView(context: Context) -> WKWebView {
        let config = WKWebViewConfiguration()
        let webView = WKWebView(frame: .zero, configuration: config)
        webView.isOpaque = false
        webView.backgroundColor = UIColor(red: 0.043, green: 0.047, blue: 0.063, alpha: 1)
        webView.scrollView.isScrollEnabled = false
        webView.scrollView.bounces = false
        onWebViewCreated?(webView)
        webView.loadHTMLString(buildHTML(side: side, highlighted: highlightedMuscles), baseURL: nil)
        return webView
    }

    func updateUIView(_ webView: WKWebView, context: Context) {
        // Re-load when side changes; re-highlight when muscles change
        webView.loadHTMLString(buildHTML(side: side, highlighted: highlightedMuscles), baseURL: nil)
    }

    // MARK: - HTML builder

    private func buildHTML(side: MuscleMapSide, highlighted: [String]) -> String {
        let svg = side == .front ? frontSVG() : backSVG()
        let highlightedJSON = highlighted
            .map { "\"\($0.lowercased())\"" }
            .joined(separator: ",")

        return """
        <!DOCTYPE html>
        <html>
        <head>
        <meta name="viewport" content="width=device-width,initial-scale=1,user-scalable=no">
        <style>
          * { box-sizing: border-box; margin: 0; padding: 0; }
          html, body {
            width: 100%; height: 100%;
            background: \(bgHex);
            display: flex; align-items: center; justify-content: center;
            overflow: hidden;
          }
          svg { width: 90%; max-height: 90vh; }
          .muscle-group { transition: fill 0.25s ease; cursor: default; }
          .muscle-label {
            font-family: -apple-system, sans-serif;
            font-size: 9px;
            fill: \(labelHex);
            pointer-events: none;
            user-select: none;
          }
        </style>
        </head>
        <body>
        \(svg)
        <script>
        (function() {
          var highlighted = [\(highlightedJSON)];
          var accent  = "\(accentHex)";
          var def     = "\(defaultHex)";

          document.querySelectorAll(".muscle-group").forEach(function(el) {
            var muscle = el.getAttribute("data-muscle");
            el.style.fill = highlighted.includes(muscle) ? accent : def;
          });
        })();
        </script>
        </body>
        </html>
        """
    }

    // MARK: - Front SVG

    // Diagrammatic representation — 12 muscle groups laid out as labelled shapes.
    // All paths carry data-muscle attributes matching the public muscle name strings.
    private func frontSVG() -> String {
        """
        <svg viewBox="0 0 160 340" xmlns="http://www.w3.org/2000/svg">
          <!-- Traps (upper) -->
          <rect class="muscle-group" data-muscle="traps"
                x="56" y="28" width="48" height="16" rx="4"/>
          <text class="muscle-label" x="80" y="39" text-anchor="middle">Traps</text>

          <!-- Shoulders -->
          <rect class="muscle-group" data-muscle="shoulders"
                x="26" y="42" width="22" height="22" rx="6"/>
          <text class="muscle-label" x="37" y="56" text-anchor="middle">Delt</text>

          <rect class="muscle-group" data-muscle="shoulders"
                x="112" y="42" width="22" height="22" rx="6"/>
          <text class="muscle-label" x="123" y="56" text-anchor="middle">Delt</text>

          <!-- Chest -->
          <rect class="muscle-group" data-muscle="chest"
                x="52" y="44" width="56" height="32" rx="6"/>
          <text class="muscle-label" x="80" y="63" text-anchor="middle">Chest</text>

          <!-- Biceps -->
          <rect class="muscle-group" data-muscle="biceps"
                x="26" y="68" width="18" height="38" rx="6"/>
          <text class="muscle-label" x="35" y="90" text-anchor="middle">Bi</text>

          <rect class="muscle-group" data-muscle="biceps"
                x="116" y="68" width="18" height="38" rx="6"/>
          <text class="muscle-label" x="125" y="90" text-anchor="middle">Bi</text>

          <!-- Abs -->
          <rect class="muscle-group" data-muscle="abs"
                x="58" y="80" width="44" height="48" rx="5"/>
          <text class="muscle-label" x="80" y="107" text-anchor="middle">Abs</text>

          <!-- Forearms (front, lower arm) -->
          <rect class="muscle-group" data-muscle="forearms"
                x="22" y="110" width="18" height="30" rx="5"/>
          <text class="muscle-label" x="31" y="128" text-anchor="middle">FA</text>

          <rect class="muscle-group" data-muscle="forearms"
                x="120" y="110" width="18" height="30" rx="5"/>
          <text class="muscle-label" x="129" y="128" text-anchor="middle">FA</text>

          <!-- Hip flexors / Hip -->
          <rect class="muscle-group" data-muscle="hip flexors"
                x="58" y="132" width="44" height="18" rx="4"/>
          <text class="muscle-label" x="80" y="144" text-anchor="middle">Hip Flex</text>

          <!-- Quads -->
          <rect class="muscle-group" data-muscle="quads"
                x="54" y="154" width="20" height="70" rx="6"/>
          <text class="muscle-label" x="64" y="192" text-anchor="middle">Quad</text>

          <rect class="muscle-group" data-muscle="quads"
                x="86" y="154" width="20" height="70" rx="6"/>
          <text class="muscle-label" x="96" y="192" text-anchor="middle">Quad</text>

          <!-- Calves (front view — tibialis) -->
          <rect class="muscle-group" data-muscle="calves"
                x="56" y="230" width="18" height="50" rx="5"/>
          <text class="muscle-label" x="65" y="258" text-anchor="middle">Calf</text>

          <rect class="muscle-group" data-muscle="calves"
                x="86" y="230" width="18" height="50" rx="5"/>
          <text class="muscle-label" x="95" y="258" text-anchor="middle">Calf</text>

          <!-- Body outline silhouette (non-interactive) -->
          <ellipse cx="80" cy="16" rx="14" ry="14"
                   fill="#222330" stroke="#2E3044" stroke-width="1"/>
          <text class="muscle-label" x="80" y="20" text-anchor="middle">Head</text>
        </svg>
        """
    }

    // MARK: - Back SVG

    private func backSVG() -> String {
        """
        <svg viewBox="0 0 160 340" xmlns="http://www.w3.org/2000/svg">
          <!-- Traps -->
          <rect class="muscle-group" data-muscle="traps"
                x="56" y="28" width="48" height="22" rx="4"/>
          <text class="muscle-label" x="80" y="42" text-anchor="middle">Traps</text>

          <!-- Rear Delts / Shoulders -->
          <rect class="muscle-group" data-muscle="shoulders"
                x="26" y="42" width="22" height="22" rx="6"/>
          <text class="muscle-label" x="37" y="56" text-anchor="middle">Delt</text>

          <rect class="muscle-group" data-muscle="shoulders"
                x="112" y="42" width="22" height="22" rx="6"/>
          <text class="muscle-label" x="123" y="56" text-anchor="middle">Delt</text>

          <!-- Lats -->
          <rect class="muscle-group" data-muscle="lats"
                x="48" y="54" width="24" height="50" rx="6"/>
          <text class="muscle-label" x="60" y="81" text-anchor="middle">Lat</text>

          <rect class="muscle-group" data-muscle="lats"
                x="88" y="54" width="24" height="50" rx="6"/>
          <text class="muscle-label" x="100" y="81" text-anchor="middle">Lat</text>

          <!-- Triceps -->
          <rect class="muscle-group" data-muscle="triceps"
                x="26" y="68" width="18" height="38" rx="6"/>
          <text class="muscle-label" x="35" y="90" text-anchor="middle">Tri</text>

          <rect class="muscle-group" data-muscle="triceps"
                x="116" y="68" width="18" height="38" rx="6"/>
          <text class="muscle-label" x="125" y="90" text-anchor="middle">Tri</text>

          <!-- Lower Back -->
          <rect class="muscle-group" data-muscle="lower back"
                x="60" y="108" width="40" height="28" rx="5"/>
          <text class="muscle-label" x="80" y="125" text-anchor="middle">Lower Back</text>

          <!-- Glutes -->
          <rect class="muscle-group" data-muscle="glutes"
                x="54" y="140" width="20" height="32" rx="6"/>
          <text class="muscle-label" x="64" y="159" text-anchor="middle">Glute</text>

          <rect class="muscle-group" data-muscle="glutes"
                x="86" y="140" width="20" height="32" rx="6"/>
          <text class="muscle-label" x="96" y="159" text-anchor="middle">Glute</text>

          <!-- Hamstrings -->
          <rect class="muscle-group" data-muscle="hamstrings"
                x="54" y="176" width="20" height="50" rx="6"/>
          <text class="muscle-label" x="64" y="203" text-anchor="middle">Ham</text>

          <rect class="muscle-group" data-muscle="hamstrings"
                x="86" y="176" width="20" height="50" rx="6"/>
          <text class="muscle-label" x="96" y="203" text-anchor="middle">Ham</text>

          <!-- Calves -->
          <rect class="muscle-group" data-muscle="calves"
                x="56" y="230" width="18" height="50" rx="5"/>
          <text class="muscle-label" x="65" y="258" text-anchor="middle">Calf</text>

          <rect class="muscle-group" data-muscle="calves"
                x="86" y="230" width="18" height="50" rx="5"/>
          <text class="muscle-label" x="95" y="258" text-anchor="middle">Calf</text>

          <!-- Head -->
          <ellipse cx="80" cy="16" rx="14" ry="14"
                   fill="#222330" stroke="#2E3044" stroke-width="1"/>
          <text class="muscle-label" x="80" y="20" text-anchor="middle">Head</text>
        </svg>
        """
    }
}
